org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareLauncherProxy.java

```diff
@@ -35,6 +46,45 @@ private ConnectionAwareLauncherProxy(ProjectConnection connection, LongRunningOp
         this.launcher = target;
     }
 
+    @SuppressWarnings("unchecked")
+    static <T> ModelBuilder<T> newModelBuilder(ProjectConnection connection, Class<T> model) {
+        ModelBuilder<T> builder = connection.model(model);
+        return (ModelBuilder<T>) newProxyInstance(connection, builder);
+    }
+
+    @SuppressWarnings("unchecked")
+    static <T> BuildActionExecuter<Collection<T>> newCompositeModelQueryExecuter(ProjectConnection connection, Class<T> model) {
+        BuildActionExecuter<Collection<T>> executer = connection.action(compositeModelQuery(model));
+        return (BuildActionExecuter<Collection<T>>) newProxyInstance(connection, executer);
+    }
+
+    private static <T> BuildAction<Collection<T>> compositeModelQuery(Class<T> model) {
+        if (Platform.inDevelopmentMode()) {
```

Comment:
See my comments in the `ideFriendlyCompositeModelQuery` method.

Before it was not a problem as the build action class was always loaded from the tooling-commons jar.